rfcs/0024-Non-consensus-node-protocol.md

```diff
@@ -0,0 +1,224 @@
+
+### Staged-ledger-validity
+
+When a node verifies a blockchain snark, it also verifies the ledger state upto a certain sequence of transactions (all the transactions for which there is a ledger proof). The transactions that haven't been completely snarked yet, full nodes apply them to the ledger in staged ledger and verify the staged-ledger-hash that is part of the new block is equal to the one they generated. This however is not possible for non-consensus nodes because they don't have any state apart from the two different block headers. We can make this slightly better for the non-consensus nodes by having the block porducers commit to the transactions being included and having an in-snark check for preventing any possible transaction reversal
+
+This can be done by chaining the ledger hash (of the ledger in the staged ledger) like we currently do for coinbases in the pending coinbase data structure.
+
+The stack operations would now be:
+
+    ``` ocaml
+        module Stack : sig
+            val push : t -> Coinbase.t option -> Frozen_ledger_hash.t -> t
+
+            module Checked : sig
+                val push : t -> Coinbase_data.var option -> Frozen_ledger_hash.var -> (t, 'a) Tick.Checked.t
+            end
+        end
+    ```
+The blockchain snark would then prove that the transactions included in a new ledger proof is infact what is expected by checking that the stack corresponding to the new ledger proof is the same as the one updated in the pending coinbase.
+
+### Shared Libraries
+
+Binary for non-consensus nodes can share the following libraries with the one for full-node:
+
+    1. Gossip net
+    2. Verifier (This needs fixing as it currently loads prover keys as well)
+    3. Snarky libraries (update and reuse snarkette?)
+    4. Best tip prover
+    5. Network pool (for transactions)
+    6. `select` function from the consensus library
+    7. Account, Transaction, user command, blockchain state, consensus state and all its contents 
+
+### Incentivizing-full-nodes
+
+The aim is to run non-consensus nodes natively, on browsers and on mobile platforms using much less resources compared to full nodes. This makes it much more accessible for users to interact with the blockchain. However, non-consensus nodes don't participate in the consensus and full nodes have no direct incentive to gossip block headers or respond to other requests-
+
+    1. Gossip block headers of root and the best tip along with the merkle-list of state hashes. These can be generated without any extra computational cost though
+    2. Respond to merkle-path requests for every new block and new root
+
+So, we need to think about ways to incentivize full nodes. Maybe-
+
+ 1. Pay a full node for fulfilling a request?
+ 2. Some sort of periodic fee for being a non-consensus node by associating it with an account?
```

Comment:
I doubt either of these are possible. For 1, the cost of fulfilling a request of this type is so low that it'd be less than transaction fees and there's no way to prove that a node did or didn't actually fulfill a request so payment agreements can't be enforced externally. For 2, I'd expect any scheme to be vulnerable to Sybil attacks.